akka-stream/src/main/scala/akka/stream/impl/fusing/GraphStages.scala

```diff
@@ -117,6 +120,51 @@ object GraphStages {
     }
   }
 
+  private class TerminationWatcher[T] extends GraphStageWithMaterializedValue[FlowShape[T, T], Future[Unit]] {
+    val name = "TerminationWatcher"
+    val in = Inlet[T](name + ".in")
+    val out = Outlet[T](name + ".out")
+    override def initialAttributes = Attributes.name(name)
+    override val shape = FlowShape(in, out)
+
+    override def createLogicAndMaterializedValue(inheritedAttributes: Attributes): (GraphStageLogic, Future[Unit]) = {
+      trait SharingTerminationFuture {
```

Comment:
Ah, thanks, I expanded the wrong code section.

I'd do this instead:

~~~scala

      val finishPromise = Promise[Unit]
      val logic = new GraphStageLogic(shape) with SharingTerminationFuture {

        setHandler(in, new InHandler {
          override def onPush(): Unit = push(out, grab(in))

          override def onUpstreamFinish(): Unit = {
            finishPromise.success(())
            super.onUpstreamFinish()
          }

          override def onUpstreamFailure(ex: Throwable): Unit = {
            finishPromise.failure(ex)
            super.onUpstreamFailure(ex)
          }
        })
        setHandler(out, new OutHandler {
          override def onPull(): Unit = pull(in)
          override def onDownstreamFinish(): Unit = {
            finishPromise.success(())
            super.onDownstreamFinish()
          }
        })
      }
      (logic, finishPromise.future)
~~~